### PR TITLE
[MEV] refactors to allow reordering of transactions by gas price

### PR DESCRIPTION
**Pull request by laura-makdah**
moves consensus index to the subdag instead of on each cert, and updates the ExecutionState `handle_consensus_transaction` to take the entire subdag instead of individual transactions, and in doing this creating the ExecutionIndices for the first time on the Sui side  

In subsequent PRs we will: 
-  reorder all the transactions within all batches of the subdag at deserialization time
-  simplify the execution indices further to only include the index of the subdag

### CODE REVIEW

**Comment by sblackshear** [comment]
A thought: would it make sense to sort by something like `(!tx.has_shared_objects, tx.gas_price)` rather than `tx.gas_price`? 

This is basically setting up two fee markets: one for single-owner txes and one for shared. At some point, we'll have an unbounded number of fee markets object-specific based on shared object hotness and will want to do something more nuanced than this. But in the short term, the scheme above will prioritize all single-writer txes (which are parallel execution friendly + on-chain MEV-free by construction) over all shared object ones, which is (I think) what we want for maximizing throughput/fee revenue.

**Comment by andll** [comment]
> reorder all the transactions within each batch

I have high level question about this PR. It seems that what we do here is essentially two things
(1) We significantly change nw executor interface to pass entire subdag to sui
(2) We add MEV logic that only reorder transactions within the batch(according to PR description)

If we only care about order of transactions within the batch, do you think we could significantly reduce blast radius of this PR by simply passing `Vec<Transaction>` to sui handler from the notifier? It seems that we could achieve MEV goal this way and the PR would be 3x smaller and easier to review?

And then if at some point there will be some separate motivation to pass entire subdag perhaps we can make a separate PR for that?

**Comment by laura-makdah** [comment]
@andll The goal is reordering all the transactions in each subdag. 
This change does touch a lot of code, but ultimately allows a lot of major simplifications and optimizations which I am glad to take advantage of. Some interesting things to consider:
1. The ordering that was created before in consensus came from arbitrary enumeration, and so constructing execution indices there was arbitrary complexity. The execution indices can be simplified further, this is still in progress. 
2. It removes the need for an entire NW component
3. This change also allows us to avoid needing to update db indices to disk to the on every transaction execution, at the tradeoff of iterating through the subdag and checking each transaction's execution status on recovery.

**Comment by andll** [comment]
@laura-makdah just to clarify - are you planning to sort transactions by gas price only inside the batch (in which case I don't understand why do we need to bundle all those nice changes in one PR)? Or are you planning to sort them across entire subdag(in which case, perhaps PR description need an update, since it seem to state sorting is with the batch?)

**Comment by laura-makdah** [comment]
@andll description updated, we are taking all transaction in the commit, irrespective of the cert & batch, and reordering them all in one big pile

File: narwhal/executor/src/subscriber.rs [hunk 1]
```diff
@@ -164,65 +164,68 @@ impl<Network: SubscriberNetwork> Subscriber<Network> {
...
 }
 
+async fn run_notify<State: ExecutionState + Send + Sync + 'static>(
+    state: State,
+    mut tr_notify: metered_channel::Receiver<ConsensusOutput>,
+) {
+    loop {
+        while let Some(message) = tr_notify.recv().await {
+            let sub_dag = message.sub_dag.clone();
+            state.notify_commit_boundary(&sub_dag).await;
```
Comment (andll) [inline narwhal/executor/src/subscriber.rs:174]: I think at this point it is reasonable to remove `notify_commit_boundary` method from the ExecutorState.

In the ConsensusHandler you can call `Authority::notify_commit_boundary` at the end of the dag processing

File: crates/sui-core/src/consensus_handler.rs [hunk 2]
```diff
@@ -63,58 +63,74 @@ impl ConsensusHandler {
...
-        let index = Self::update_hash(
-            &self.last_seen,
-            consensus_index.clone(),
-            &serialized_transaction,
-        );
-        let index = if let Some(index) = index {
-            index
-        } else {
-            debug!(
+        let mut sequenced_transactions = Vec::new();
```
Comment (andll) [inline crates/sui-core/src/consensus_handler.rs:71]: Do you mind adding few utilization timers here?(you can see examples how `utilization_timer` is used in authority.rs for example)

Specifically -
(a) consensus_task_utilization for overall duration of `handle_consensus_output` function.
(b) When we have sorting, can we move that to separate function and have utilization timer for measuring sorting specifically

File: narwhal/executor/src/subscriber.rs [hunk 3]
```diff
@@ -55,47 +52,52 @@ struct Fetcher<Network> {
...
+    // This is ugly but has to be done this way for now
+    // Currently network incorporate both server and client side of RPC interface
+    // To construct server side we need to set up routes first, which requires starting Primary
+    // Some cleanup is needed
+
+    let (tx_notifier, rx_notifier) =
+        metered_channel::channel(primary::CHANNEL_CAPACITY, &metrics.tx_notifier);
+
+    vec![
+        spawn_monitored_task!(async move { run_notify(state, rx_notifier).await }),
```
Comment (andll) [inline narwhal/executor/src/subscriber.rs:75]: I believe you can simply write `spawn_monitored_task!(run_notify(state, rx_notifier))` here

File: narwhal/executor/src/subscriber.rs [hunk 4]
```diff
@@ -55,47 +52,52 @@ struct Fetcher<Network> {
...
+        spawn_monitored_task!(async move {
```
Comment (andll) [inline narwhal/executor/src/subscriber.rs:76]: Do you mind putting all that `async` code into separate method, perhaps `Subscriber::start`?

And then this line can become just `spawn_monitored_task!(Subscriber::start(...))`

File: narwhal/types/src/primary.rs [hunk 5]
```diff
@@ -699,7 +699,7 @@ impl Certificate {
     Ord,
     Arbitrary,
 )]
-pub struct CertificateDigest([u8; crypto::DIGEST_LENGTH]);
+pub struct CertificateDigest(pub(crate) [u8; crypto::DIGEST_LENGTH]);
```
Comment (andll) [inline narwhal/types/src/primary.rs:702]: Let's just add some accessor method to CertificateDigest,
Something like `into_bytes(self) -> [u8; crypto::DIGEST_LENGTH]` if we need to extract value, or `as_bytes(&self) -> &[u8; crypto::DIGEST_LENGTH]` if we need a reference

File: crates/sui-core/src/consensus_handler.rs [hunk 6]
```diff
@@ -63,58 +63,74 @@ impl ConsensusHandler {
...
+                            continue;
+                        }
+                    };
+                    let index = ExecutionIndices {
+                        last_committed_round: round,
+                        sub_dag_index: consensus_output.sub_dag.sub_dag_index,
+                        transaction_index: seq,
+                    };
+
+                    let index_with_hash =
```
Comment (andll) [inline crates/sui-core/src/consensus_handler.rs:98]: You slightly change order between deserialization and update_hash here - previously we always called update_hash before deserialization, and in this implementation you skip `update_hash` if deserialization fails. I think original approach we always calculating hash makes more sense, so perhaps you can move it before deserialization?

File: crates/sui-core/src/consensus_handler.rs [hunk 7]
```diff
@@ -63,58 +63,74 @@ impl ConsensusHandler {
...
-            &serialized_transaction,
-        );
-        let index = if let Some(index) = index {
-            index
-        } else {
-            debug!(
+        let mut sequenced_transactions = Vec::new();
+        let mut seq = 0;
+
+        for (cert, batches) in consensus_output.batches {
```
Comment (andll) [inline crates/sui-core/src/consensus_handler.rs:74]: Note that `Notifier` had some metrics such as `notifier_processed_batches`, `batch_execution_latency` and `notifier_processed_bytes`. Unless I missed something it feels like we just abandoned them in this PR. Perhaps you can move them from narwhal to sui and populate them here?

**Review by andll — CHANGES_REQUESTED** [review]
This PR looks pretty good, great job!

In terms of next steps, after you take care of comments, given the size of this change, do you mind deploying this PR to private testnet and check some metrics to make sure everything works as expected? Perhaps check narwhal round is incrementing, as well as `last_certified_checkpoint` metric to confirm checkpoints are working. And you can check your newly added `consensus_task_utilization` metric to see where we are right now.

Also, I think PR title need change now that it is limited to consensus interface refactoring only [and by the way, thank you for separating this refactoring in one PR and doing MEV reordering separately, this makes review easier].

File: narwhal/executor/src/lib.rs [hunk 8]
```diff
@@ -42,12 +41,7 @@ pub type SerializedTransactionDigest = u64;
 // Important - if you add method with the default implementation here make sure to update impl ExecutionState for Arc<T>
 pub trait ExecutionState {
```
Comment (andll) [inline narwhal/executor/src/lib.rs:42]: At this point we can change `ExecutionState::load_execution_indices` to only return u64 with `last_committed_round` right? And ExecutionIndexes at this point can be entirely moved to sui

File: narwhal/consensus/src/tusk.rs [hunk 9]
```diff
@@ -96,33 +93,27 @@ impl ConsensusProtocol for Tusk {
...
-            committed_sub_dags.push(CommittedSubDag {
+            let sub_dag = CommittedSubDag {
                 certificates: sequence,
                 leader: leader.clone(),
-            });
+                sub_dag_index: consensus_index,
+            };
+
+            // Increase the global consensus index.
+            consensus_index += 1;
```
Comment (andll) [inline narwhal/consensus/src/tusk.rs:109]: Btw, I am not sure what is the need for a global consensus index at this point?
For execution indexes should we just have something like `{commit_round, certificate_index_in_commit, batch_index_in_certificate, txn_index_in_batch}`?

Comment (asonnino) [reply narwhal/consensus/src/tusk.rs:109]: I am not sure it is needed at this point, but removing it may be easier on another PR because I think it will require changing the crash-recovery strategy

Comment (asonnino) [reply narwhal/consensus/src/tusk.rs:109]: I am also not sure we need the consensus store anymore

File: narwhal/executor/src/subscriber.rs (hunk 4 above)
Comment (laura-makdah) [reply narwhal/executor/src/subscriber.rs:76]: The reason why we spawn 2 monitored tasks is so that the two can be done in parallel and not just concurrently. Do you know if there is a way to make it parallel with just one task? I didn't think there was.

Comment (andll) [reply narwhal/executor/src/subscriber.rs:76]: We should spawn two tasks, I am just asking to move contents of `async` block into separate method

File: crates/sui-core/src/consensus_handler.rs (hunk 6 above)
Comment (laura-makdah) [reply crates/sui-core/src/consensus_handler.rs:98]: The reason I moved it to be after the deserialization is that right after the deserialization, we will next reorder the transactions, and so the re-hashing should subsequently follow the execution order, which is only formed now after deserialization. We can add a call to update the hash in the error branch of deserialization? Can you describe a bit more about what that would be used for?

Comment (andll) [reply crates/sui-core/src/consensus_handler.rs:98]: Ok, I think I understand the problem.

The hash is used to detect consensus forks (we have already found two bugs in narwhal that cause forks about a month ago thanks to that hash). As such it does not need to follow execution order, it is ok to just calculate hash based on the narwhal output, regardless on which order transactions are executed.

Ok, I guess in this case no need to make code more complicated then needed, we can change the logic the way you do it in this PR

File: crates/sui-core/src/consensus_handler.rs (hunk 7 above)
Comment (laura-makdah) [reply crates/sui-core/src/consensus_handler.rs:74]: Ah good point there, since the notifier no longer does any processing I considered them obsolete. The meanings will be a bit different, so maybe we create equivalent ones with separate names. After the reordering of transactions, we won't be executing batches sequentially anymore for example.

Comment (andll) [reply crates/sui-core/src/consensus_handler.rs:74]: We can have different names, but I think measuring those things are still important.
The batch reordering will matter for `batch_execution_latency` metric only right? I think @akichidis added it, so perhaps he can add context on what should replace it with new architecture. Perhaps it can just measure time to when batch is deserialized in handler?

File: narwhal/executor/src/subscriber.rs (hunk 4 above)
Comment (laura-makdah) [reply narwhal/executor/src/subscriber.rs:76]: ah I see

File: narwhal/consensus/src/tusk.rs (hunk 9 above)
Comment (laura-makdah) [reply narwhal/consensus/src/tusk.rs:109]: Yes, I am planning to address all these in the next PR 👍

File: narwhal/executor/src/subscriber.rs [hunk 10]
```diff
@@ -250,6 +273,9 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
             stagger += Duration::from_millis(200);
         }
         let (batch, _, _) = futures::future::select_all(futures).await;
+        self.metrics
+            .batch_execution_latency
```
Comment (laura-makdah) [inline narwhal/executor/src/subscriber.rs:277]: Oh I actually did move this one over @andll

File: narwhal/executor/src/subscriber.rs [hunk 11]
```diff
@@ -55,47 +52,85 @@ struct Fetcher<Network> {
...
-    })
+        )),
+    ]
+}
+
+async fn run_notify<State: ExecutionState + Send + Sync + 'static>(
+    state: State,
+    mut tr_notify: metered_channel::Receiver<ConsensusOutput>,
+) {
+    loop {
```
Comment (akichidis) [inline narwhal/executor/src/subscriber.rs:94]: Is this `loop` needed here? I believe the `while` on its own should be enough

Comment (akichidis) [reply narwhal/executor/src/subscriber.rs:94]: Also I get that this task will leave for as long the `tx/tr_notify` channels will live, but I would still suggest making it listen to the `tx_reconfigure` for any shutdown messages - let's keep things explicit

Comment (laura-makdah) [reply narwhal/executor/src/subscriber.rs:98]: ✔️

File: crates/sui-core/src/consensus_handler.rs [hunk 12]
```diff
@@ -63,85 +64,94 @@ impl ConsensusHandler {
...
-        let index = if let Some(index) = index {
-            index
-        } else {
-            debug!(
+        let _scope = monitored_scope("HandleConsensusOutputFull");
+        let mut sequenced_transactions = Vec::new();
+        let mut seq = 0;
+
+        for (cert, batches) in consensus_output.batches {
+            let round = cert.header.round;
```
Comment (andll) [inline crates/sui-core/src/consensus_handler.rs:77]: This actually needs to be a sub_dag round, not certificate round. `CommittedSubDag::round` can be used for that, and you can evaluate it outside of this loop

File: crates/sui-core/src/consensus_handler.rs [hunk 13]
```diff
@@ -63,85 +64,94 @@ impl ConsensusHandler {
...
-            consensus_index.clone(),
-            &serialized_transaction,
-        );
-        let index = if let Some(index) = index {
-            index
-        } else {
-            debug!(
+        let _scope = monitored_scope("HandleConsensusOutputFull");
+        let mut sequenced_transactions = Vec::new();
+        let mut seq = 0;
```
Comment (andll) [inline crates/sui-core/src/consensus_handler.rs:74]: For consistency with how it was implemented previously, perhaps you should move this `seq` inside the loop below, so that it will be index in the certificate

**Review by andll — CHANGES_REQUESTED** [review]

File: crates/sui-core/src/consensus_handler.rs (hunk 13 above)
Comment (laura-makdah) [reply crates/sui-core/src/consensus_handler.rs:74]: Now the sequence is assigned for every transaction in the sub_dag, rather than for every transaction in the cert. This sequence generation is going to move to a separate function in the next PR.

File: crates/sui-core/src/consensus_handler.rs (hunk 12 above)
Comment (andll) [reply crates/sui-core/src/consensus_handler.rs:77]: Oh yeah - cert.header.round is the round of given certificate, while CommittedSubDag::round is the round of leader certificate of the sub dag, subdag can have certificates from different rounds

File: crates/sui-core/src/authority/authority_store.rs [hunk 14]
```diff
@@ -1309,7 +1309,6 @@ impl<S: Eq + Debug + Serialize + for<'de> Deserialize<'de>> SuiDataStore<S> {
             next_versions,
         )
     }
-
```
Comment (andll) [inline crates/sui-core/src/authority/authority_store.rs:1312]: nit: revert this change(in most of code we have a newline between functions)

File: crates/sui-core/src/consensus_handler.rs [hunk 15]
```diff
@@ -63,89 +65,104 @@ impl ConsensusHandler {
...
+                        }
+                    };
+                    let index = ExecutionIndices {
+                        last_committed_round: round,
+                        sub_dag_index: consensus_output.sub_dag.sub_dag_index,
+                        transaction_index: seq,
+                    };
+
+                    let index_with_hash =
+                        match Self::update_hash(&self.last_seen, index, &transaction.tracking_id) {
```
Comment (andll) [inline crates/sui-core/src/consensus_handler.rs:102]: I don't think it's a good idea to use tracking_id for integrity hash - there is no any guarantees about how other party calculates it. Let's just use serialized transaction content as before

File: narwhal/consensus/src/bullshark.rs [hunk 16]
```diff
@@ -158,39 +153,29 @@ impl ConsensusProtocol for Bullshark {
...
             };
+
+            // Increase the global consensus index.
+            state.latest_sub_dag_index = next_sub_dag_index;
+
+            // Persist the update.
+            // TODO [issue #116]: Ensure this is not a performance bottleneck.
             self.store
-                .write_committed_sub_dag(&state.last_committed, &sub_dag)?;
+                .write_consensus_state(&state.last_committed, &sub_dag)?;
```
Comment (andll) [inline narwhal/consensus/src/bullshark.rs:176]: Because you have `latest_sub_dag_index` in memory, you should maintain consistency between what is in DB and in memory. This consistency can be violated if method here returns error. (in memory state is updated and db is not)

You can fix it just by swapping two lines and updating in-memory state after db write succeed

File: narwhal/executor/src/lib.rs [hunk 17]
```diff
@@ -117,42 +96,32 @@ pub async fn get_restored_consensus_output<State: ExecutionState>(
...
+    // that need to be sent for execution.
 
-    let last_committed_leader = execution_state
-        .load_execution_indices()
-        .await
-        .last_committed_round;
+    let last_executed_sub_dag_index = execution_state.last_executed_sub_dag_index().await;
 
     let compressed_sub_dags =
-        consensus_store.read_committed_sub_dags_from(&last_committed_leader)?;
```
Comment (andll) [inline narwhal/executor/src/lib.rs:130]: Perhaps we can rename `compressed_sub_dag(s)` -> `sub_dag(s)` here and in the loop below, since we no longer compress subdugs. (Which btw I liked, since it simplified consensus recovery and it is much easier to reason about it right now)

**Review by andll — COMMENTED** [review]
Overall this looks pretty good now!

I see that `notifier_processed_batches` and `notifier_processed_bytes` metrics are still exist and are not used, can you move them to sui and populate? (And remove from subscriber metrics on nw side)

File: narwhal/consensus/src/bullshark.rs (hunk 16 above)
Comment (laura-makdah) [reply narwhal/consensus/src/bullshark.rs:176]: good catch

File: crates/sui-core/src/consensus_handler.rs (hunk 15 above)
Comment (laura-makdah) [reply crates/sui-core/src/consensus_handler.rs:102]: I see, ok

**Review by andll — APPROVED** [review]
Thank you @laura-makdah, this is epic!

**Comment by laura-makdah** [comment]
@andll thank you for the extremely helpful review!
